benchmark/benchmark.hpp

```diff
@@ -0,0 +1,326 @@
+/*
+ * Copyright (C) 2024 Intel Corporation
+ *
+ * Under the Apache License v2.0 with LLVM Exceptions. See LICENSE.TXT.
+ * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+ *
+ */
+
+#include <benchmark/benchmark.h>
+#include <random>
+#include <umf/memory_pool.h>
+#include <umf/memory_provider.h>
+
+#include "benchmark_interfaces.hpp"
+
+struct alloc_data {
+    void *ptr;
+    size_t size;
+};
+
+#define UMF_BENCHMARK_TEMPLATE_DEFINE(BaseClass, Method, ...)                  \
+    BENCHMARK_TEMPLATE_DEFINE_F(BaseClass, Method, __VA_ARGS__)                \
+    (benchmark::State & state) {                                               \
+        for (auto _ : state) {                                                 \
+            bench(state);                                                      \
+        }                                                                      \
+    }
+
+#define UMF_BENCHMARK_REGISTER_F(BaseClass, Method)                            \
+    BENCHMARK_REGISTER_F(BaseClass, Method)                                    \
+        ->ArgNames(                                                            \
+            BENCHMARK_PRIVATE_CONCAT_NAME(BaseClass, Method)::argsName())      \
+        ->Name(BENCHMARK_PRIVATE_CONCAT_NAME(BaseClass, Method)::name())       \
+        ->MinWarmUpTime(1)
+
+class fixed_alloc_size : public alloc_size_interface {
+  public:
+    unsigned SetUp(::benchmark::State &state, unsigned r) override {
+        size = state.range(r);
+        return r;
+    }
+    void TearDown([[maybe_unused]] ::benchmark::State &state) override {}
+    size_t nextSize() override { return size; };
+    static std::vector<std::string> argsName() { return {"size"}; }
+
+  private:
+    size_t size;
+};
+
+class uniform_alloc_size : public alloc_size_interface {
+    using distribution = std::uniform_int_distribution<int64_t>;
+
+  public:
+    unsigned SetUp(::benchmark::State &state, unsigned r) override {
+        auto min = state.range(r);
+        auto max = state.range(r + 1);
```

Comment:
r + 1/2? comment what this function does

Reply:
done